File: Utility/Browser/clsWeb.vb
Imports System
Imports System.IO
Imports System.Net
Imports System.Collections.Generic
Imports System.Collections.ObjectModel
Imports OpenQA.Selenium
Imports OpenQA.Selenium.PhantomJS
Imports HtmlAgilityPack
Imports System.Runtime.Serialization.Formatters.Binary

Public Class clsWeb
    Implements IDisposable

    Private _driver As IWebDriver
    Private _options As PhantomJSOptions
    Private _service As PhantomJSDriverService
    Private _cookie As Boolean = False
    'Private UserAgent As String = "Chrome/34.0.1847.131"

    Sub New()

        _service = PhantomJSDriverService. _
                   CreateDefaultService(Environment.CurrentDirectory)
        _service.HideCommandPromptWindow = True

        _options = New PhantomJSOptions
        '_options.AddAdditionalCapability("phantomjs.page.settings.userAgent", UserAgent)

        _driver = New PhantomJSDriver(_service, _options)
        _driver.Manage().Timeouts().ImplicitlyWait(New TimeSpan(0, 0, 30))

    End Sub

    Sub Dispose() Implements IDisposable.Dispose

        _driver.Quit()
        _driver = Nothing
        _options = Nothing
        _service = Nothing

    End Sub

    Private Function InnerElements(ctrl As IWebElement) As Boolean

        Dim ret As Boolean = False
        Dim html As String
        Dim doc As HtmlDocument = Nothing

        html = ctrl.GetAttribute("outerHTML")

        doc = CleanHTML(html)

        If doc.DocumentNode.FirstChild.HasChildNodes Then
            If doc.DocumentNode.FirstChild.FirstChild.Name = "#text" Then
                ret = False
            Else
                ret = True
            End If
        End If

        doc = Nothing

        Return ret

    End Function

    Private Function ProcessElement( _
                                     ByRef item As IWebElement, _
                                     ByRef kvp As KeyValuePair(Of String, String) _
                                   ) As Boolean

        Dim ret As Boolean = False
        Dim check As String

        check = item.GetAttribute("name")
        If check = kvp.Key Then
            Select Case item.GetAttribute("type")
                Case "text", "password"
                    item.SendKeys(kvp.Value)
                Case "checkbox"
                    item.Click()
            End Select
            ret = True
        End If

        Return ret

    End Function

    Private Sub GetInnerElements( _
                                  ByRef ctrl As IWebElement, _
                                  ByRef Fields As List(Of KeyValuePair(Of String, String)) _
                                )

        Dim idx As Integer
        Dim ret As Boolean

        Dim kvp As KeyValuePair(Of String, String)
        Dim ChildElements As ReadOnlyCollection(Of IWebElement)

        ChildElements = ctrl.FindElements(By.XPath("*"))

        For Each item As IWebElement In ChildElements

            If InnerElements(item) Then
                GetInnerElements(item, Fields)
            End If

            For idx = 0 To (Fields.Count - 1)

                kvp = Fields(idx)
                ret = ProcessElement(item, kvp)

                If ret = True Then
                    Exit For
                End If

            Next

        Next

    End Sub

    Public Sub FollowLink(ByVal URL As String, ByVal LinkText As String, cookieName As String)

        Dim cookies As ReadOnlyCollection(Of OpenQA.Selenium.Cookie)
        Dim cookies2 As New List(Of OpenQA.Selenium.Cookie)
        Dim link As IWebElement
        Dim ret As String

        _driver.Navigate.GoToUrl(URL)

        link = _driver.FindElement(By.LinkText(LinkText))

        link.Click()

        ret = _driver.PageSource

        cookies = _driver.Manage.Cookies.AllCookies

        For idx = 0 To cookies.Count - 1
            cookies2.Add(cookies(idx))
        Next

        WriteCookiesToDisk(cookieName, cookies2)

    End Sub

    Public Sub LogIn(ByVal URL As String, formName As String, Fields As List(Of KeyValuePair(Of String, String)), cookieName As String)

        Dim cookies As ReadOnlyCollection(Of OpenQA.Selenium.Cookie)
        Dim cookies2 As New List(Of OpenQA.Selenium.Cookie)

        Dim idx As Integer
        Dim check As String
        Dim ret As Boolean
        Dim kvp As KeyValuePair(Of String, String)
        Dim formElement As IWebElement
        Dim allFormChildElements As ReadOnlyCollection(Of IWebElement)

        Dim btn As IWebElement = Nothing

        _driver.Navigate.GoToUrl(URL)

        formElement = _driver.FindElement(By.Id(formName))
        allFormChildElements = formElement.FindElements(By.XPath("*"))

        For Each item As IWebElement In allFormChildElements

            If InnerElements(item) Then
                GetInnerElements(item, Fields)
            End If

            For idx = 0 To (Fields.Count - 1)

                kvp = Fields(idx)

                ret = ProcessElement(item, kvp)

                If ret = True Then
                    Exit For
                End If

                check = item.GetAttribute("type")
                If check = kvp.Key Then
                    If item.Text = kvp.Value Then
                        btn = item
                        Exit For
                    End If
                End If

            Next

        Next

        btn.Submit()

        cookies = _driver.Manage.Cookies.AllCookies

        For idx = 0 To cookies.Count - 1
            cookies2.Add(cookies(idx))
        Next

        WriteCookiesToDisk(cookieName, cookies2)

    End Sub

    Public Function DownloadPage(ByVal URL As String, Optional ByVal Cookie As String = "") As String

        Dim cookies As List(Of OpenQA.Selenium.Cookie)
        Dim idx As Integer

        Dim ret As String = ""

        _driver.Navigate.GoToUrl(URL)

        If System.IO.File.Exists(Environment.CurrentDirectory & "\" & Cookie) Then

            cookies = ReadCookiesFromDisk(Cookie, URL)
            _driver.Manage.Cookies.DeleteAllCookies()

            For idx = 0 To cookies.Count - 1
                _driver.Manage.Cookies.AddCookie(cookies(idx))
            Next

            _driver.Navigate.GoToUrl(URL)

            _cookie = True

        End If

        ret = _driver.PageSource

        Return ret

    End Function

    Private Sub WriteCookiesToDisk(ByVal fileName As String, ByVal cookieJar As List(Of OpenQA.Selenium.Cookie))

        Dim cookies As New CookieContainer
        Dim cookie As System.Net.Cookie

        Dim idx As Integer

        For idx = 0 To cookieJar.Count - 1
            cookie = New System.Net.Cookie( _
                                            cookieJar(idx).Name, _
                                            cookieJar(idx).Value, _
                                            cookieJar(idx).Path, _
                                            cookieJar(idx).Domain _
                                          )

            cookie.Expires = cookieJar(idx).Expiry

            cookies.Add(cookie)

        Next

        Using stream As Stream = File.Create(fileName)
            Try
                'MsgBox("Writing cookies to disk... ")
                Dim formatter As New BinaryFormatter()
                formatter.Serialize(stream, cookies)
                'MsgBox("Done.")
            Catch e As Exception
                MsgBox("Problem writing cookies to disk: " & e.Message)
            End Try
        End Using
    End Sub

    Private Function ReadCookiesFromDisk(ByVal fileName As String, URL As String) As List(Of OpenQA.Selenium.Cookie)

        Dim cookies As CookieContainer
        Dim cookiec As CookieCollection

        Dim ts As New TimeSpan

        Dim dte As DateTime = DateTime.Now

        dte = dte.AddDays(365)

        Dim u As New System.Uri(URL)

        URL = u.Scheme & Uri.SchemeDelimiter & GetDomain(u)

        u = Nothing
        u = New System.Uri(URL)

        Dim idx As Integer

        Dim cookiejar As New List(Of OpenQA.Selenium.Cookie)
        Dim cookie As OpenQA.Selenium.Cookie

        Try
            Using stream As Stream = File.Open(fileName, FileMode.Open)
                'MsgBox("Reading cookies from disk... ")
                Dim formatter As New BinaryFormatter()
                'MsgBox("Done.")
                cookies = CType(formatter.Deserialize(stream), CookieContainer)
                cookiec = cookies.GetCookies(u)

                For idx = 0 To cookiec.Count - 1
                    cookie = New OpenQA.Selenium.Cookie( _
                                                         cookiec(idx).Name, _
                                                         cookiec(idx).Value, _
                                                         cookiec(idx).Domain, _
                                                         cookiec(idx).Path, _
                                                         dte _
                                                       )

                    cookiejar.Add(cookie)

                Next


            End Using
        Catch e As Exception
            MsgBox("Problem reading cookies from disk: " & e.Message())
            cookiejar = Nothing
        End Try

        Return cookiejar

    End Function

    Private Function GetDomain(u As Uri) As String
        If u.HostNameType <> UriHostNameType.Dns Then Return String.Empty
        Dim parts = u.Host.Split("."c)
        If parts.Length > 1 Then
            Return String.Join(".", parts, parts.Length - 2, 2)
        Else
            Return parts(0)
        End If
    End Function

End Class

File: FFDownload/HtmlScraper/Utility/Browser/clsWeb.cs
using System.Collections.ObjectModel;
using System.Net;
using System.Runtime.Serialization.Formatters.Binary;
using HtmlScraper.Utility.Browser.HTML;
using Microsoft.VisualBasic;
using OpenQA.Selenium;
using OpenQA.Selenium.Support.UI;
using OpenQA.Selenium.Firefox;
using NHibernate.Driver;

namespace HtmlScraper.Utility.Browser
{

    public class clsWeb : IDisposable
    {

        private IWebDriver _driver;
        private FirefoxOptions _options;
        private FirefoxDriverService _service;
        private bool _cookie = false;
        private IWebElement _submit;

        // Private UserAgent As String = "Chrome/34.0.1847.131"

        public clsWeb()
        {

            _service = FirefoxDriverService.CreateDefaultService(Environment.CurrentDirectory);

            _service.HideCommandPromptWindow = true;

            _options = new FirefoxOptions();
            // _options.AddAdditionalCapability("phantomjs.page.settings.userAgent", UserAgent)

            _driver = new FirefoxDriver(_service, _options);
            WebDriverWait wait = new WebDriverWait(_driver, TimeSpan.FromSeconds(30));

        }

        public void Dispose()
        {

            _driver.Quit();
            _driver = null;
            _options = null;
            _service = null;

        }

        private bool InnerElements(IWebElement ctrl)
        {

            bool ret = false;
            string html;
            HtmlAgilityPack.HtmlDocument doc = null;

            html = ctrl.GetAttribute("outerHTML");

            doc = modHTML.CleanHTML(ref html);

            if (doc.DocumentNode.FirstChild.HasChildNodes)
            {

                if (doc.DocumentNode.FirstChild.ChildNodes.Count > 1)
                {
                    ret = true;
                }
                else if (doc.DocumentNode.FirstChild.FirstChild.Name == "#text")
                {
                    ret = false;
                }
                else
                {
                    ret = true;
                }
            }

            doc = null;

            return ret;

        }

        private bool CheckTag(string name)
        {

            switch (name ?? "")
            {
                case "input":
                case "select":
                    {
                        return true;
                    }

                default:
                    {
                        return false;
                    }
            }


        }

        private void ProcessFields(IWebElement item, ref List<KeyValuePair<string, string>> Fields)


        {

            int idx;
            KeyValuePair<string, string> kvp;
            bool ret;

            if (CheckTag(item.TagName))
            {

                var loopTo = Fields.Count - 1;
                for (idx = 0; idx <= loopTo; idx++)
                {

                    kvp = Fields[idx];
                    ret = ProcessElement(ref item, ref kvp);

                    if (ret == true)
                    {
                        break;
                    }

                }

            }

        }

        private bool ProcessElement(ref IWebElement item, ref KeyValuePair<string, string> kvp)


        {

            SelectElement sel;
            bool ret = false;
            string name;
            string type;

            name = item.GetAttribute("name");
            type = item.GetAttribute("type");

            if ((name ?? "") == (kvp.Key ?? ""))
            {
                switch (type ?? "")
                {
                    case "text":
                    case "password":
                        {
                            item.SendKeys(kvp.Value);
                            break;
                        }
                    case "checkbox":
                        {
                            item.Click();
                            break;
                        }
                    case "select-one":
                        {
                            sel = new SelectElement(item);
                            sel.SelectByValue(kvp.Value);
                            break;
                        }
                }
                ret = true;
            }

            if (!ret)
            {
                if ((type ?? "") == (kvp.Key ?? ""))
                {
                    if ((item.Text ?? "") == (kvp.Value ?? "") | (item.GetAttribute("value") ?? "") == (kvp.Value ?? ""))
                    {
                        _submit = item;
                    }
                }
            }


            return ret;

        }

        private IWebElement? GetInnerElements(IWebElement? ctrl, ref List<KeyValuePair<string, string>> Fields)
        {

            if (ctrl == null)
                return null;

            IWebElement? ret = null;

            ReadOnlyCollection<IWebElement> ChildElements;

            ChildElements = ctrl.FindElements(By.XPath("*"));

            foreach (IWebElement item in ChildElements)
            {

                if (InnerElements(item))
                {
                    ret = GetInnerElements(item, ref Fields);
                }

                ProcessFields(ret, ref Fields);

            }

            return null;

        }

        public void FollowLink(string URL, string LinkText, string cookieName)
        {

            ReadOnlyCollection<OpenQA.Selenium.Cookie> cookies;
            var cookies2 = new List<OpenQA.Selenium.Cookie>();
            IWebElement link;
            string ret;

            _driver.Navigate().GoToUrl(URL);

            link = _driver.FindElement(By.LinkText(LinkText));

            link.Click();

            ret = _driver.PageSource;

            cookies = _driver.Manage().Cookies.AllCookies;

            for (int idx = 0, loopTo = cookies.Count - 1; idx <= loopTo; idx++)
                cookies2.Add(cookies[idx]);

            WriteCookiesToDisk(cookieName, cookies2);

        }

        public void LogIn(string URL, string formName, List<KeyValuePair<string, string>> Fields, string cookieName)
        {

            ReadOnlyCollection<OpenQA.Selenium.Cookie> cookies;
            var cookies2 = new List<OpenQA.Selenium.Cookie>();

            int idx;
            IWebElement formElement;
            ReadOnlyCollection<IWebElement> forms;
            ReadOnlyCollection<IWebElement> allFormChildElements;

            IWebElement btn = null;

            _driver.Navigate().GoToUrl(URL);

            forms = _driver.FindElements(By.TagName("form"));

            formElement = forms[0];

            allFormChildElements = formElement.FindElements(By.XPath("*"));

            IWebElement? ret = null;

            foreach (IWebElement item in allFormChildElements)
            {

                if (InnerElements(item))
                {
                    ret = GetInnerElements(item, ref Fields);
                }

                ProcessFields(ret, ref Fields);

            }

            _submit.Submit();

            cookies = _driver.Manage().Cookies.AllCookies;

            var loopTo = cookies.Count - 1;
            for (idx = 0; idx <= loopTo; idx++)
                cookies2.Add(cookies[idx]);

            WriteCookiesToDisk(cookieName, cookies2);

        }

        public string DownloadPage(string URL, string Cookie = "")
        {

            List<OpenQA.Selenium.Cookie> cookies;
            int idx;

            string ret = "";

            _driver.Navigate().GoToUrl(URL);

            if (File.Exists(Environment.CurrentDirectory + @"\" + Cookie))
            {

                cookies = ReadCookiesFromDisk(Cookie, URL);
                _driver.Manage().Cookies.DeleteAllCookies();

                var loopTo = cookies.Count - 1;
                for (idx = 0; idx <= loopTo; idx++)
                    _driver.Manage().Cookies.AddCookie(cookies[idx]);

                _driver.Navigate().GoToUrl(URL);

                _cookie = true;

            }

            ret = _driver.PageSource;

            return ret;

        }

        private void WriteCookiesToDisk(string fileName, List<OpenQA.Selenium.Cookie> cookieJar)
        {

            var cookies = new CookieContainer();
            System.Net.Cookie cookie;

            int idx;

            var loopTo = cookieJar.Count - 1;
            for (idx = 0; idx <= loopTo; idx++)
            {
                cookie = new System.Net.Cookie(cookieJar[idx].Name, cookieJar[idx].Value, cookieJar[idx].Path, cookieJar[idx].Domain) { Expires = (DateTime)cookieJar[idx].Expiry };

                cookies.Add(cookie);

            }

            using (Stream stream = File.Create(fileName))
            {
                try
                {
                    // MsgBox("Writing cookies to disk... ")
                    var formatter = new BinaryFormatter();
                    formatter.Serialize(stream, cookies);
                }
                // MsgBox("Done.")
                catch (Exception e)
                {
                    Interaction.MsgBox("Problem writing cookies to disk: " + e.Message);
                }
            }
        }

        private List<OpenQA.Selenium.Cookie> ReadCookiesFromDisk(string fileName, string URL)
        {

            CookieContainer cookies;
            CookieCollection cookiec;

            var ts = new TimeSpan();

            var dte = DateTime.Now;

            dte = dte.AddDays(365d);

            var u = new Uri(URL);

            // URL = u.Scheme & Uri.SchemeDelimiter & GetDomain(u)

            // u = Nothing
            // u = New System.Uri(URL)

            int idx;

            var cookiejar = new List<OpenQA.Selenium.Cookie>();
            OpenQA.Selenium.Cookie cookie;

            try
            {
                using (Stream stream = File.Open(fileName, FileMode.Open))
                {
                    // MsgBox("Reading cookies from disk... ")
                    var formatter = new BinaryFormatter();
                    // MsgBox("Done.")
                    cookies = (CookieContainer)formatter.Deserialize(stream);
                    cookiec = cookies.GetCookies(u);

                    var loopTo = cookiec.Count - 1;
                    for (idx = 0; idx <= loopTo; idx++)
                    {
                        cookie = new OpenQA.Selenium.Cookie(cookiec[idx].Name, cookiec[idx].Value, cookiec[idx].Domain, cookiec[idx].Path, dte);






                        cookiejar.Add(cookie);

                    }


                }
            }
            catch (Exception e)
            {
                Interaction.MsgBox("Problem reading cookies from disk: " + e.Message);
                cookiejar = null;
            }

            return cookiejar;

        }

        private string GetDomain(Uri u)
        {
            if (u.HostNameType != UriHostNameType.Dns)
                return string.Empty;
            var parts = u.Host.Split('.');
            if (parts.Length > 1)
            {
                return string.Join(".", parts, parts.Length - 2, 2);
            }
            else
            {
                return parts[0];
            }
        }

    }
}